### DISCUSSION TITLE
Features Whishlist

### DISCUSSION THREAD

**Discussion by MarcoRavich** [FORARTfe/hALSAmrec#1] [Ideas]
Buffer size formula (needed ?):
- @anselbrandt's [pi-recorder capture](https://github.com/anselbrandt/pi-recorder/blob/master/capture.js));

NTFS / ExFAT / EXT4 partitioning (if needed):
- @azurekite83's [partition-aid](https://github.com/azurekite83/partition-aid)
- @jas- [bashAutopartition ](https://github.com/jas-/bashAutopartition)
- @mohsenkamini's [Partitioning-Automation](https://github.com/mohsenkamini/Partitioning-Automation);
- @toniles' [BASH-Dual-Boot-USB-Creator-with-Data-Storage](https://github.com/toniles/BASH-Dual-Boot-USB-Creator-with-Data-Storage);
- @jnoliv's [mkusb-part](https://github.com/jnoliv/mkusb-part);
- @markonisic's [shrink-split-sda](https://github.com/markonisic/shrink-split-sda);

More to come.

**Comment by MarcoRavich** [comment]
Interesting (ChatGPT-generated) scripts by @ergz: [cl-audio-recorder](https://github.com/ergz/cl-audio-recorder)

**Comment by mohsenkamini** [comment]
Happy to help if you needed anything!

**Comment by jonneymendoza** [comment]
What's this for?

**Comment by MarcoRavich** [comment]
Hi there, 1st of all thanks for reply.

The hALSAmrec goal is to headlessly record raw multichannels from (pro) audio interfaces by just connecting them into a SBC (I'm using a RaspBerry Pi 3B+ now).

The inspiration came from @bfields's [openwrt-recording](https://github.com/bfields/openwrt-recording/) script but the idea is to "generalize" and enrich it to let any - adequate - SBC:
- autocreate user-needed recording partition (Win = NTFS, MacOS = ExFAT, Lin = Ext4) on SD if needed;
- autoprobe (hotplug) connected-interface characteristics;
- autostart recording of all channels.

So the aim is to make enduser have to:
- install his preferred (DE/GUI-less) distro;
- install the (interactive, e.g. asking the recording-partition type) script;
- connect-and-record.

However, since I'm a Bash-noob, I'm collecting resources - and therefore possible collaborators - here.

Thanks in advance for any support you can provide.

**Comment by MarcoRavich** [comment]
_Safe System Shutdown_ from @chechocardona's [AutoRecording](https://github.com/chechocardona/AutoRecording)

_SSH Custom Buttons_ from @bassmanitram's [XR18-Raspberry-Pi-Tools](https://github.com/bassmanitram/XR18-Raspberry-Pi-Tools#readme)

**Comment by knutopia** [comment]
@MarcoRavich for what it's worth, I had modlogr working fine, but there would be occasional drop outs after hours of recording, where a second of audio or so would randomly be missing... possibly related to raspberry pi system activity or hardware driver issues. Who knows. Watch out for those.

**Reply by MarcoRavich** [reply]
> @MarcoRavich for what it's worth, I had modlogr working fine, but there would be occasional drop outs after hours of recording, where a second of audio or so would randomly be missing... possibly related to raspberry pi system activity or hardware driver issues. Who knows. Watch out for those.

Hi there, thanks for reply.

I checked - and starred - [Modlogr](https://github.com/knutopia/Modlogr#readme) and I really apreciate many aspects of it (such as Intallation process, LED management, etc).
The goal of hALSAmrec is to have as many features as possible (one for all: start/stop control through SSH commands generated by an external device - e.g. smartphone/tablet - connected in WiFi-direct to the Raspberry) but based on the "[KISS principle](https://en.wikipedia.org/wiki/KISS_principle)".
So the aim is to perform everything exploiting already available - or installing fewer additional packages as possible - tools in minimal RBPi distros (OpenWRT - at the moment - in my case); that's why, for example, it favor ALSA commands over FFMPEG/SOX/etc.

I'm bit busy these days (working on [another customization project](https://github.com/MarcoRavich/cctv-viewer)), anyway I'm trying to implement the auto audio interface probing (as stated [here](https://github.com/bfields/openwrt-recording/issues/3) too) using this code:
```
`# arecord absolute path
ARECORD="/usr/bin/arecord"

# Probing interface
$ARECORD -D hw:0,0 --dump-hw-params > /tmp/arecord_output.txt 2>&1
arecord_output=$(cat /tmp/arecord_output.txt)

max_channels=$(echo "$arecord_output" | grep "CHANNELS:" | sed -n 's/.*CHANNELS: \[\?\([0-9]*\)\( \([0-9]*\)\)\?\]/\3/p')
if [ -z "$max_channels" ]; then
    max_channels=$(echo "$arecord_output" | grep "CHANNELS:" | sed -n 's/.*CHANNELS: \([0-9]*\)/\1/p')
fi

bitformat=$(echo "$arecord_output" | grep "^FORMAT:" | sed -n 's/.*FORMAT: \(.*\)/\1/p' | awk '{print $NF}')
if [ -z "$bitformat" ]; then
    bitformat=$(echo "$arecord_output" | grep "^FORMAT:" | sed -n 's/.*FORMAT: \([A-Z0-9_]*\)/\1/p')
fi

extract_max_from_range() {
    echo "$1" | sed -n 's/.*[[(]\([0-9]*\) \([0-9]*\)[])]/\2/p'
}

max_buffertime=$(echo "$arecord_output" | grep "BUFFER_TIME:" | while read -r line; do extract_max_from_range "$line"; done)
max_buffersize=$(echo "$arecord_output" | grep "BUFFER_SIZE:" | while read -r line; do extract_max_from_range "$line"; done)

max_rate=$(echo "$arecord_output" | grep "RATE:" | while read -r line; do extract_max_from_range "$line"; done)
if [ -z "$max_rate" ]; then
    max_rate=$(echo "$arecord_output" | grep "RATE:" | sed -n 's/.*RATE: \([0-9]*\)/\1/p')
fi`
```

It works, but breaks the "hotplug" function...

**Comment by MarcoRavich** [comment]
Ok, an optimized distro - like @nooploop's [PieJam OS](https://github.com/nooploop/piejam_os#readme) - would be cool.
